src/sharing.rs

```diff
@@ -1,35 +1,159 @@
 #![cfg(feature = "share_clipboard")]
 
 #[cfg(feature = "share_selection")]
-use clipboard::x11_clipboard::{Primary, X11ClipboardContext};
+use clipboard::{x11_clipboard, x11_clipboard::X11ClipboardContext};
+#[cfg(feature = "share_selection")]
+use wl_clipboard_rs::utils::{is_primary_selection_supported, PrimarySelectionCheckError};
+
 #[cfg(not(feature = "share_selection"))]
 use clipboard::ClipboardContext;
 use clipboard::ClipboardProvider;
+use std::io::Read;
+use wl_clipboard_rs::{
+    copy,
+    copy::{Options, Source},
+    paste,
+    paste::{get_contents, Error, Seat},
+};
 
 #[cfg(not(feature = "share_selection"))]
 pub fn read_share() -> Option<String> {
-    ClipboardProvider::new()
-        .and_then(|mut ctx: ClipboardContext| ctx.get_contents())
-        .ok()
+    if std::option_env!("XDG_SESSION_TYPE").unwrap_or_default() == "wayland"
+        || !std::option_env!("WAYLAND_DISPLAY")
+            .unwrap_or_default()
+            .is_empty()
+        || std::option_env!("GDK_BACKEND").unwrap_or_default() == "wayland"
```

Comment:
Can this be extracted to a separate function for reuse/deduplication?

Reply:
Thanks for your suggestion! I have now added an `is_wayland` function.